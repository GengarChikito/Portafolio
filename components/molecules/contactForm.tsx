import { Form, Input, Button, message } from "antd";

/**
 * Defino las 'props' para este componente.
 * En este caso, no necesito ninguna prop, pero mantengo la interfaz
 * por si decido añadirle algo en el futuro (como un título).
 */
export interface ContactFormProps {}

/**
 * Este es mi componente 'ContactForm', una "molécula".
 * Es un formulario simple de Ant Design para que los visitantes
 * me puedan enviar un mensaje.
 */
export function ContactForm(_: ContactFormProps) {
    // Uso el hook 'useForm' de Ant Design.
    // Esto me da una instancia de 'form' que puedo usar para
    // controlar el formulario, como por ejemplo, para limpiarlo.
    const [form] = Form.useForm();

    /**
     * Esta es la función que se ejecuta cuando el formulario se envía
     * y *pasa* todas las validaciones (es decir, todos los campos
     * requeridos están llenos).
     */
    async function onFinish(): Promise<void> {
        // Por ahora, no estoy enviando el formulario a un backend (es un TODO).
        // Simplemente muestro un mensaje de éxito con 'message.success'.
        message.success("Gracias por tu mensaje. Te responderé pronto.");

        // Uso la instancia 'form' para limpiar (resetear) los campos
        // después de que se envía exitosamente.
        form.resetFields();
    }

    return (
       /* // Aquí empieza el formulario.
        // layout="vertical" pone las etiquetas (label) encima de los campos.
        // Le paso mi instancia 'form' y mi función 'onFinish'.*/
        <Form layout="vertical" form={form} onFinish={onFinish}>

            {/* Campo 1: Nombre */}
            <Form.Item
                label="Nombre"
                name="name"
               /* // 'rules' define las reglas de validación.
                // Aquí, es un campo obligatorio.*/
                rules={[{ required: true, message: "Ingresa tu nombre" }]}
            >
                <Input placeholder="Tu nombre" />
            </Form.Item>

            {/* Campo 2: Correo */}
            <Form.Item
                label="Correo"
                name="email"
                // Añado dos reglas: es obligatorio y debe ser de tipo 'email'.
                rules={[{ required: true, type: "email", message: "Correo válido" }]}
            >
                <Input placeholder="tu@correo.com" />
            </Form.Item>

            {/* Campo 3: Mensaje */}
            <Form.Item
                label="Mensaje"
                name="msg"
                rules={[{ required: true, message: "Escribe tu mensaje" }]}
            >
                {/* Uso 'Input. TextArea' para un campo de texto más grande. */}
                <Input.TextArea rows={5} placeholder="¿En qué puedo ayudarte?" />
            </Form.Item>

            {/* Botón de envío */}
            <Button type="primary" htmlType="submit">
                Enviar
            </Button>
        </Form>
    );
}