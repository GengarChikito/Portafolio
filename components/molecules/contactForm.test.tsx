import React from "react";
import { describe, it, expect, vi } from "vitest";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { ContactForm } from "./contactForm";


// üß© Mock de antd
const formSpy = vi.fn();
const messageSpy = vi.fn();

const mockFormInstance = {
    resetFields: vi.fn(),
};

vi.mock("antd", () => {
    const Input = (props: any) => (
        <input data-testid={props["data-testid"] || "input"} {...props} />
    );
    Input.TextArea = (props: any) => (
        <textarea data-testid="textarea" {...props}></textarea>
    );

    return {
        Form: Object.assign(
            ({ children, onFinish, layout, form }: any) => {
                formSpy({ layout, form });
                // simulamos el submit manualmente con un bot√≥n
                return (
                    <form
                        data-testid="form"
                        onSubmit={(e) => {
                            e.preventDefault();
                            onFinish?.();
                        }}
                    >
                        {children}
                    </form>
                );
            },
            {
                useForm: () => [mockFormInstance],
                Item: ({ children, label }: any) => (
                    <label data-testid={`label-${label}`}>
                        {label}
                        {children}
                    </label>
                ),
            }
        ),
        Input,
        Button: ({ children, htmlType }: any) => (
            <button data-testid="button" type={htmlType}>
                {children}
            </button>
        ),
        message: {
            success: (msg: string) => messageSpy(msg),
        },
    };
});


describe("contactForm", () => {
    it("renderiza los campos y etiquetas correctamente", () => {
        render(<ContactForm />);
        expect(screen.getByTestId("label-Nombre")).toBeInTheDocument();
        expect(screen.getByTestId("label-Correo")).toBeInTheDocument();
        expect(screen.getByTestId("label-Mensaje")).toBeInTheDocument();

        // Inputs visibles
        expect(screen.getAllByRole("textbox").length).toBeGreaterThan(0);
        expect(screen.getByTestId("textarea")).toBeInTheDocument();
    });

    it("al enviar el formulario muestra mensaje de √©xito y resetea campos", async () => {
        render(<ContactForm />);

        const form = screen.getByTestId("form");
        fireEvent.submit(form);

        await waitFor(() => {
            expect(messageSpy).toHaveBeenCalledWith(
                "Gracias por tu mensaje. Te responder√© pronto."
            );
        });

        expect(mockFormInstance.resetFields).toHaveBeenCalledTimes(1);
    });

    it("snapshot del formulario", () => {
        const { container } = render(<ContactForm />);
        expect(container).toMatchSnapshot();
    });
});
