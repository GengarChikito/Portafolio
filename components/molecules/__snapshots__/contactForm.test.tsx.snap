// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`contactForm > snapshot del formulario 1`] = `
<div>
  <form
    data-testid="form"
  >
    <label
      data-testid="label-Nombre"
    >
      Nombre
      <input
        data-testid="input"
        placeholder="Tu nombre"
      />
    </label>
    <label
      data-testid="label-Correo"
    >
      Correo
      <input
        data-testid="input"
        placeholder="tu@correo.com"
      />
    </label>
    <label
      data-testid="label-Mensaje"
    >
      Mensaje
      <textarea
        data-testid="textarea"
        placeholder="¿En qué puedo ayudarte?"
        rows="5"
      />
    </label>
    <button
      data-testid="button"
      type="submit"
    >
      Enviar
    </button>
  </form>
</div>
`;
